doc/proposals/ansible-operator.md

```diff
@@ -0,0 +1,101 @@
+  * Will contain a wrapper type that conforms to the sdk.Handler, for a user to use with standard operator-sdk workflows. The wrapper type has default methods for everything, but they are overridable.
+* /operator-sdk/pkg/ansible/runner
+  * Contains the runner types and interfaces
+  * Implementation is behind an internal package (/operator-sdk/pkg/ansible/runner/internal)
+  * New Methods are exposed and are the main way a user interacts with the package
+  * Runner interface for running ansible from the operator.
+  * NewForWatchers - the method that returns a map of GVK to Runner types based on the watchers file.
+  * NewPlaybookRunner - the method that returns a new Runner for a playbook.
+  * NewRoleRunner - the method that returns a new Runner for a role.
+  * This contains the events API code and public methods. Implementation should probably be in the internal package. The events API is used for recieving events from ansible runner. 
+
+* /operator-sdk/pkg/ansible/proxy
+  * This is a reverse proxy for the kubernetes API that is used for owner reference injection.
+* /operator-sdk/pkg/ansible/proxy/kubeconfig
+  * Code needed to generate the kubeconfig for the proxy.
+
+
+### Commands
+We are adding and updating existing commands to accommodate the ansible operator.  Changes to the `cmd` package as well as changes to the generator are needed. 
+
+`operator-sdk new --type ansible --kind <kind> --api-version <group/version>`  This will be a new generation command under the hood. We will:
+* Create a new ansible role in the roles directory
+* Create a new watchers file. The role and GVK are defaulted based on input to the command. 
+* A CRD is generated. This CRD does not have any validations defined. 
+* A dockerfile is created using the watchers file and the ansible role with the base image being the ansible operator base image.
+
+The resulting structure should be
+```
+|- Dockerfile
+|
+|- roles
+\ | - <kind>
+|  \ | - generated ansible role
+|
+| - watchers.yaml
+|
+| - deploy
+\  | - <kind>-CRD.yaml
+|  | - rbac.yaml
+|  | - operator.yaml
+|  | - cr.yaml
+```
+
+`operator-sdk generate crd <api-version> <kind> ` This will be used to generate new CRDs based on ansible code. The command helps when a user wants to watch more than one CRD for their operator.
+Args:
+Required kind - the kind for the object.  
+Required api-version - the <group>/<version> for the CRD.
+Flags:
+Optional: --defaults-file - A path to the defaults file to use to generate a new CRD.yaml. If this is not defined, then an empty CRD is created.
+
+`operator-sdk up ansible` - This should use the known structure and the ansible operator code to run the operator from this location. The command works by running the operator-sdk binary, which includes the ansible operator code, as the operator process. This is slightly different than the current up local command. For debugging a user could then set a flag --verbose which tells ansible to be more verbose.
```

Comment:
This needs a bit thought. Currently, we have `operator-sdk up local` to run test for go operator. Maybe we can have `operator-sdk up local --type=ansible` to run ansible operator locally.

Reply:
So I have two thoughts

1. I don't want ansible operator developers to feel like a second-class citizen from the commands. The constant need to add type arguments worries me in this sense because we are not giving them a "default" way of running their code.

2. The idea of a "remote" command seems like a good idea. Your right that we would want the ansible operator to be used for that command as well. 

I wonder if we could do something like:

`operator-sdk up {local,remore}`  - where the presence of a `cmd` directory with a `.go` file somewhere inside means we `go run`, while the absence means we use ansible. 

The only issue I worry about with this is a possible poor user experience if their project is in a very odd state. It might be hard for us to determine errors where if they are explicit we can do more error checking. 

Thoughts?

Reply:
magic detection can cause all sorts of problems. Also, assuming one when its not the other also doesn't scale when a 3rd option comes around. I'm for a little more typing that is always consistent. don't make go lang a special thing and just add a required, type flag = [go, ansible]

Reply:
@fanminshi Are you ok with this? I think I agree with @kfox1111

Reply:
@shawn-hurley  @kfox1111 I think we default the `--type` flag to `go` for backward compatibility. However, if user can use  `--type go` if they want to. In essence, we should have a `--type` for all supported types but with the default to `go`. That's my opinion on this. If we want to make the flag required, that can cause backward compatibility issue which is also something we can discuss about. I am not for breaking backward compatibility in general. Since the project is still in early stage, I am fine with it if everyone thinks it is a good idea.